### DISCUSSION TITLE
Doc: VSCode use case (prevents commits)

### DISCUSSION THREAD

**Discussion by bbros-dev** [evilmartians/lefthook#361] [Q&A]
This is likely a documentation issue.
When using Lefthook Git hooks the commit flow is 'broken by the output generated by Lefthook (see below).
The consequence is that the commit is not made and the user is prompted with a dialog showing the first line of output captured `Lefthook v0.7.1`.

Appreciate any hints or tips on how to setup/use Lefthook with VSCode.

The VSCode `OUPUT` window:

```bash
> git commit --quiet --allow-empty-message --file - --signoff
Lefthook v0.7.1
RUNNING HOOKS GROUP: pre-commit

  EXECUTE > terraform_docs

  EXECUTE > terraform_fmt

  EXECUTE > terraform_tflint

  EXECUTE > terraform_tfsec

  EXECUTE > terraform_validate

  EXECUTE > terragrunt_fmt

  EXECUTE > terragrunt_validate

SUMMARY: (done in 0.15 seconds)
🥊  terraform_docs
🥊  terraform_fmt
🥊  terraform_tflint
🥊  terraform_tfsec
🥊  terraform_validate
🥊  terragrunt_fmt
🥊  terragrunt_validate
> git config --get-all user.name
> git config --get-all user.email
> git ls-files --stage -- /home/<redacted>/src/kfchub.tf/lefthook.yml
> git ls-files --stage -- /home/<redacted>/src/kfchub.tf/.lefthook/pre-commit/terraform_docs
> git cat-file -s a8f62f8eabf15d3f11bc0525b15a6c5437f35b23
> git cat-file -s 92597d839a4a9c493568e81dbef6db8b328ffd94
> git show --textconv :lefthook.yml
> git show --textconv :.lefthook/pre-commit/terraform_docs
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse development
> git rev-parse --symbolic-full-name development@{u}
> git rev-list --left-right development...refs/remotes/upstream/development
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
```

**Comment by sheldonhull** [comment]
I had this issue with pre-commit as well, which really interrupted my workflow. Any solutions found for this?

**Comment by tetienne** [comment]
Any update on this topic?

**Comment by mrexox** [comment]
Hey! Did you try lefthook v1? Does the issue still persist on a new version?

**Comment by TheComputerM** [comment]
Its still an issue, I use https://biomejs.dev/recipes/git-hooks/#lefthook, if I git commit from the bash terminal, it works fine, but if I use the vscode source control view to commit, it errors out:

```
sh: line 2: npx: command not found
```

For context I use pnpm, so npm isn't installed through my package manager but pnpm's node version manager

**Reply by mrexox** [reply]
I think you can consider `rc` option: https://lefthook.dev/configuration/rc.html. It was added specifically for these kind of things

**Comment by otg996** [comment]
I don't think this is an issue with lefthook, it appears to be a PATH issue. I faced this same issue in the past with a global-to-chained-local hooks workflow, this issue arises because hooks use `sh` for execution, while your PATH is very likely extended using a `.zsh/.bashrc` file. 

It's worth noting that `.bashrc`'s first line's task is to do nothing if not running in interactive mode, because of this, I personally dislike editing PATH from a `bash` exclusive configuration file. My solution is to move PATH and any environment variables out of `.bashrc` into a terminal agnostic shell file, then source that from `.bashrc`.

The solution for this is to ensure VSCode has access to the updated path before clicking the button, you can *probably* add an entry in lefthook's config, under all hooks that you use to source the file containing your environment variables (as long as they're not in `.bashrc`).
I say probably because this is not the workflow I follow but I see no reason why it wouldn't work, it would perhaps be clumsy and unmaintainable if you have lots of hooks enabled, but it *should* work.

My custom solution involves:
- Completely disable the ability for lefthook to install itself in the global hooks path
- Define a base hook script that sources environment variables and sets up all hook prerequisites
- Source that script from all my (global) hooks
- Chain to local hooks (lefthook's hooks)

This results in VSCode's commit/push button working as expected when using lefthook.
Edits: formatting

Edit 2: mrexox mentions the rc option here, a much more straightforward way to address this issue. https://github.com/evilmartians/lefthook/discussions/885#discussioncomment-11469737
